### PR TITLE
[WiP] Update training code to RasterVision version 0.21

### PR DESCRIPTION
**Pull request by jamesmcclain**

### CODE REVIEW

File: training/pipeline.py [hunk 1]
```diff
@@ -82,21 +93,33 @@ def hrefs_to_sceneconfig(
...
+    ymax = int(extent_crop[3] * width)
+    xmax = int(extent_crop[2] * height)
+    bbox = Box(ymin=ymin, xmin=xmin, ymax=ymax, xmax=xmax)
     image_source = RasterioSourceConfig(
         uris=[imagery],
         allow_streaming=True,
         channel_order=channel_order,
         transformers=transformers,
-        extent_crop=extent_crop,
+        # bbox=bbox,  # XXX
```
Comment (jamesmcclain) [inline training/pipeline.py:112]: @AdeelH I need a little bit of help updating this repository to use RasterVision 0.21.  This line fails validation ("invalid tuple"); I am not able to understand the issue.

File: training/pipeline.py [hunk 2]
```diff
@@ -105,7 +128,7 @@ def hrefs_to_sceneconfig(
         raster_source=label_raster_source)
 
     return SceneConfig(id=name,
-                       aoi_geometries=aoi,
+                       # aoi_uris=[aoi],  # XXX
```
Comment (jamesmcclain) [inline training/pipeline.py:131]: @AdeelH Earlier version of RV had an `aoi_geometries` field that would allow you to send over a dictionary of geometry rather than just an URI.  440 of 441 chips are failing the AOI filtering stage, so I think this is only filtering by the first label (as the AOI) instead of the union of the labels (which was the behavior before).

File: training/pipeline.py [hunk 3]
```diff
@@ -264,7 +287,7 @@ def get_config(runner,
                             num_epochs=epochs,
                             batch_sz=batch_sz,
                             external_loss_def=external_loss_def,
-                            ignore_last_class='force'),
```
Comment (jamesmcclain) [inline training/pipeline.py:267]: @AdeelH this was useful for making sure that the model only had two classes instead of three.  I don't know what the current behavior is.

File: training/pipeline.py [hunk 4]
```diff
@@ -281,5 +304,4 @@ def get_config(runner,
                                       predict_chip_sz=chip_sz,
                                       chip_options=chip_options,
                                       chip_nodata_threshold=.75,
-                                      img_format='npy',
```
Comment (jamesmcclain) [inline training/pipeline.py:284]: @AdeelH Is it still possible to request that chips are stored in Numpy format?

File: training/pipeline.py (hunk 1 above)
Comment (AdeelH) [reply training/pipeline.py:112]: You have to pass in a (ymin, xmin, ymax, xmax) integer tuple. The values have to be coords and not offsets.. See [docs](https://docs.rastervision.io/en/stable/api_reference/_generated/rastervision.core.data.raster_source.rasterio_source_config.RasterioSourceConfig.html#rastervision.core.data.raster_source.rasterio_source_config.RasterioSourceConfig.bbox).

File: training/pipeline.py (hunk 4 above)
Comment (AdeelH) [reply training/pipeline.py:284]: No. It now happens automatically for non 3-channel chips.

File: training/pipeline.py (hunk 3 above)
Comment (AdeelH) [reply training/pipeline.py:267]: You can use [ignore_class_index](https://docs.rastervision.io/en/stable/api_reference/_generated/rastervision.pytorch_learner.learner_config.SolverConfig.html#rastervision.pytorch_learner.learner_config.SolverConfig.ignore_class_index) to ignore a specific class in  the loss. 

RV no longer uses this info when creating the model. To use a model with a fewer, you can specify fewer classes in the `ClassConfig`. See [this `ClassConfig` tutorial](https://docs.rastervision.io/en/stable/usage/tutorials/reading_labels.html#ClassConfig) for how to avoid creating an extra null class.

File: training/pipeline.py (hunk 2 above)
Comment (AdeelH) [reply training/pipeline.py:131]: `aoi_geometries` has been removed. You can now only specify URIs.

You can debug exactly what AOIs are being used via something like:
```py
from rastervision.core.data import RasterioCRSTransformer
from rastervision.core.data.utils import get_polygons_from_uris

crs_transformer = RasterioCRSTransformer.from_uri(img_uri)
aoi_polygons = get_polygons_from_uris(aoi_uris, crs_transformer)
```

File: training/pipeline.py (hunk 1 above)
Comment (jamesmcclain) [reply training/pipeline.py:112]: Yes, I tried passing an integer tuple as well as a `Box` object with the same result.

Comment (jamesmcclain) [reply training/pipeline.py:112]: The lines 102 through 106 above show the construction, I also tried using non-constructed numbers like `Box(0, 0, 1, 1)` and and `[0, 0, 1, 1]`.

File: training/pipeline.py (hunk 4 above)
Comment (jamesmcclain) [reply training/pipeline.py:284]: :+1:

File: training/pipeline.py (hunk 2 above)
Comment (jamesmcclain) [reply training/pipeline.py:131]: I see.  What is the easiest way to pass a constructed AOI that can't be found at a specific URI?

File: training/pipeline.py (hunk 1 above)
Comment (AdeelH) [reply training/pipeline.py:112]: I just tested the below and it did not run into any errors.
```py
from rastervision.core.data import RasterioSourceConfig

RasterioSourceConfig(uris='', bbox=[0, 0, 1, 1])
```

Can you share the full stack trace?

Comment (jamesmcclain) [reply training/pipeline.py:112]: Hmmm ... okay never mind.  I must have made some mistake: it is accepting my numbers.

File: training/pipeline.py (hunk 2 above)
Comment (jamesmcclain) [reply training/pipeline.py:131]: Is it safe to create a temporary geojson file that is only accessible during the chipping stage, or will that get me in trouble in later stages?  It seems like it should be fine (in principle) because the chips will be in existence and on disk, but I'm wondering if the file will be referenced in the pipeline and cause an issue that way.

File: training/pipeline.py (hunk 1 above)
Comment (AdeelH) [reply training/pipeline.py:112]: These don't look right, BTW:
```py
ymin = int(extent_crop[1] * width)
xmin = int(extent_crop[0] * height)
ymax = int(extent_crop[3] * width)
xmax = int(extent_crop[2] * height)
```

I think this is what you want (adapted from [here](https://github.com/azavea/raster-vision/blob/0.13/rastervision_core/rastervision/core/data/raster_source/rasterio_source.py#L187-L191)):

```py
skip_top, skip_left, skip_bottom, skip_right = extent_crop
ymin = int(h * skip_top)
xmin = int(w * skip_left)
ymax = h - int(h * skip_bottom)
xmax = w - int(w * skip_right)
bbox = (ymin, xmin, ymax, xmax)
```

File: training/pipeline.py (hunk 2 above)
Comment (AdeelH) [reply training/pipeline.py:131]: I believe the scenes are also built in the `eval` stage, so the file would have to at least be available until then.
